### ISSUE TITLE
Issues with symlinks in fish config dir

### ISSUE THREAD

**Issue by jnnkB** [jorgebucaran/fisher#613]
Thanks a lot for your work on this tool. It's really great.

I use [mackup](https://github.com/lra/mackup) for syncing my config files across different devices. Mackup therefore symlinks `$__fish_config_dir/completions`,  `$__fish_config_dir/conf.d`  `$__fish_config_dir/config.fish` and  `$__fish_config_dir/functions` to an synced folder.

When running fisher update or install I get the following type of error:

```
cp: cannot overwrite non-directory <symlink> with directory <path>
``` 

It originates in line 140. 

https://github.com/jorgebucaran/fisher/blob/dc63a83c78f92bdc4d5bb0c8949bff519f73217e/fisher.fish#L144

After I changed it to the following it worked again.

```fish
for folder in $source/{functions,conf.d,completions}
    if test (count "$folder/"*) != 0
        command cp -LRf "$folder/"* "$fisher_path/"(basename "$folder")
    end
end
```

**Comment by jorgebucaran** [comment]
> @jnnkB Mackup therefore symlinks $__fish_config_dir/completions, $__fish_config_dir/conf.d $__fish_config_dir/config.fish and $__fish_config_dir/functions to an synced folder.

Looks like a bug we've had before, so this is likely a regression. Now, I'd like to reproduce this scenario in my machine, but I'm not sure what I should do. Where should I create the symlinks? Thanks.

**Comment by jnnkB** [comment]
In my fish path completions, conf.d, config.fish and functions are symlinked. In my example to a folder a sync using syncthing.
You could create a test folder in ~, and then copy completions, conf.d, config.fish and functions from $__fish_config_dir into this folder. Then you can link each of them into $__fish_config_dir.

**Comment by jorgebucaran** [comment]
> You could create a test folder in ~, and then copy completions, conf.d, config.fish and functions from $__fish_config_dir into this folder. Then you can link each of them into $__fish_config_dir.

I did this, then tried everything, install, update, remove, etc., and coulnd't reproduce the issue.

**Comment by NotAFile** [comment]
I have this issue too. For me:

```
~> file /home/nota/.config/fish/functions
/home/nota/.config/fish/functions: symbolic link to ../../dotfiles/fish/.config/fish/functions
```

**Comment by jorgebucaran** [comment]
@NotAFile Any idea how I can reproduce this? I tried @jnnkB's idea, but found that Fisher still installs to the real `functions/` directory without issues. Could this be an OS issue? I'm on macOS.

**Comment by NotAFile** [comment]
Can reproduce it with the following in a fresh `debian` container:
```sh
apt update && apt -y install fish curl
cd
mkdir -p .config/fish/
mkdir -p lndest
ln -s ~/lndest/ .config/fish/functions
fish -c "curl -sL git.io/fisher | source && fisher install jorgebucaran/fisher"
```

Result:

```
fetching https://codeload.github.com/jorgebucaran/fisher/tar.gz/HEAD
cp: cannot overwrite non-directory '/root/.config/fish/functions' with directory '/tmp/tmp.MLSSeT5hCz/functions'
installing jorgebucaran/fisher
 +/root/.config/fish/functions/fisher.fish
 +/root/.config/fish/completions/fisher.fish
source: Error encountered while sourcing file '/root/.config/fish/functions/fisher.fish':
source: No such file or directory
installed 1 plugin/s
```

**Comment by jnnkB** [comment]
Yeah your're right. It's not problem on macOS. You can test it using

```
mkdir test
ln -s test test2
mkdir test-data
mkdir test-data/test2
touch test-data//test2/data
cp -R test-data/test2 .
```

This does work on macOS but on linux you get this error:

```
cp: cannot overwrite non-directory <symlink> with directory <path>
```

**Comment by jorgebucaran** [comment]
@jnnkB Is there a clean way to do that on Linux?

**Comment by jnnkB** [comment]
I don't really know. You saw my weird fix in issue message.

**Comment by jorgebucaran** [comment]
@jnnkB Thanks, what Linux is this?

**Comment by jorgebucaran** [comment]
Could I get a confirmation that this is the problematic line?

https://github.com/jorgebucaran/fisher/blob/dc63a83c78f92bdc4d5bb0c8949bff519f73217e/fisher.fish#L144


@NotAFile @NotAFile Can

**Comment by jorgebucaran** [comment]
Can I get a confirmation that this is the problematic line?

https://github.com/jorgebucaran/fisher/blob/dc63a83c78f92bdc4d5bb0c8949bff519f73217e/fisher.fish#L144

**Comment by NotAFile** [comment]
I am curious how exactly it behaves under macOS as the requested behaviour is kind of ambigous. Does it overwrite the symlink with a folder? Or does it place the files in the linked folder destination? (second would probably be desirable, in this specific instance...)

Either way, that line is the problem, yeah. The `cp -r $source/functions $fisher_path` will try to create a directory named `functions` inside `$fisher_path`. However, a symlink by that name already exists in the destination, so creating a `functions` directory fails, as expected when creating a directory with the same name as a non-directory.

The easiest solution is probably to `cp -r $source/$folder $fisher_path/$folder` for all three folders? Alternatively one could use something non-standard like `rsync` which is better at merging folders than `cp`.

**Comment by jorgebucaran** [comment]
@NotAFile I am curious how exactly it behaves under macOS...

Same way it behaves on Linux when the destination is not a symlink. It places the files in the destination.

**Comment by jorgebucaran** [comment]
Probably fixed by f4b32b3. Please install Fisher again and let me know if the patch works for you.

cc @NotAFile @jnnkB @aureooms

**Comment by jnnkB** [comment]
Seems to work. Thanks!
